assertsprocessor/sampler.go

```diff
@@ -0,0 +1,195 @@
+	item := Item{
+		trace:   &trace,
+		ctx:     &ctx,
+		latency: summary.latency,
+	}
+	if summary.hasError || summary.isSlow {
+		// Get the trace queue for the entity and request
+		pq := s.getTraceQueues(summary.requestKey)
+		if summary.hasError {
+			s.logger.Info("Capturing error trace",
+				zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+				zap.Float64("latency", summary.latency))
+			pq.errorQueue.push(&item)
+		} else {
+			s.logger.Info("Capturing slow trace",
+				zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+				zap.Float64("latency", summary.latency))
+			pq.slowQueue.push(&item)
+		}
+	} else if len(spanSet.rootSpans) > 0 {
+		// Capture healthy samples based on configured sampling rate
+		state, _ := s.healthySamplingState.LoadOrStore(summary.requestKey.AsString(), &periodicSamplingState{
+			lastSampleTime: 0,
+			rwMutex:        &sync.RWMutex{},
+		})
+		samplingState := state.(*periodicSamplingState)
+		s.logger.Info("Found Normal trace",
+			zap.String("request", summary.requestKey.AsString()),
+			zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+			zap.Float64("latency", summary.latency),
+			zap.Time("lastSampleTime", time.Unix(samplingState.lastSampleTime, 0)))
+		if samplingState.sample(s.config.NormalSamplingFrequencyMinutes) {
+			s.logger.Info("Capturing normal trace",
+				zap.String("traceId", summary.slowestRootSpan.TraceID().String()),
+				zap.Float64("latency", summary.latency))
+			pq := s.getTraceQueues(summary.requestKey)
+
+			// Push to the latency queue to prioritize the healthy sample too
+			pq.slowQueue.push(&item)
+		}
+	}
+}
+
+func (s *sampler) getSummary(traceId string, spanSet *resourceSpanGroup) *traceSummary {
+	summary := traceSummary{}
+	summary.hasError = false
+	summary.isSlow = false
+	maxLatency := float64(0)
+	summary.hasError = summary.hasError || spanSet.hasError()
+	entityKey := buildEntityKey(s.config, spanSet.namespace, spanSet.service)
+	for _, rootSpan := range spanSet.rootSpans {
+		request := getRequest(s.requestRegexps, rootSpan)
+		summary.isSlow = summary.isSlow || s.isSlow(spanSet.namespace, spanSet.service, rootSpan)
+		max := math.Max(maxLatency, computeLatency(rootSpan))
+		if max > maxLatency {
+			maxLatency = max
+			summary.slowestRootSpan = rootSpan
+			summary.requestKey = RequestKey{
+				entityKey: entityKey,
+				request:   request,
+			}
+			summary.latency = maxLatency
+		}
+	}
+	s.logger.Info("Trace summary",
+		zap.String("traceId", traceId),
+		zap.String("request", summary.requestKey.AsString()),
+		zap.Bool("slow", summary.isSlow),
+		zap.Bool("error", summary.hasError),
+		zap.Float64("latency", summary.latency))
+	return &summary
+}
+
+func (s *sampler) getTraceQueues(key RequestKey) *traceQueues {
+	var pq, _ = s.topTracesMap.LoadOrStore(key.AsString(), &traceQueues{
+		slowQueue:  NewTraceQueue(int(math.Min(5, float64(s.config.MaxTracesPerMinutePerContainer)))),
```

Comment:
A container could be reporting traces with multiple URIs. This config is really MaxTracesPerMinutePerRequestContext than Conatiner

Reply:
> A container could be reporting traces with multiple URIs. This config is really MaxTracesPerMinutePerRequestContext than Conatiner

I thought about it, but it was getting too long. May be we can drop `MaxTraces` and change it to `LimitPerRequestPerContainer`